packages/auth/src/client/service.ts

```diff
@@ -166,14 +276,126 @@ function verifyClientDisplay(
   )
 }
 
-function verifyJwk(jwk: JWKWithRequired, keys: RegistryKey): boolean {
-  // TODO: update this to reflect eventual shape of response from registry
+function isJwkViable(keys: RegistryKey): boolean {
   return !!(
-    !keys.revoked &&
-    keys.exp &&
-    new Date() < new Date(keys.exp * 1000) &&
-    keys.nbf &&
-    new Date() >= new Date(keys.nbf * 1000) &&
-    keys.x === jwk.x
+    (!keys.exp || new Date() < new Date(keys.exp * 1000)) &&
+    (!keys.nbf || new Date() >= new Date(keys.nbf * 1000))
   )
 }
+
+function verifyJwk(jwk: JWKWithRequired, keys: RegistryKey): boolean {
+  // TODO: update this to reflect eventual shape of response from registry
+  return !!(!keys.revoked && isJwkViable(keys) && keys.x === jwk.x)
+}
+
+async function tokenHttpsigMiddleware(
+  deps: ServiceDependencies,
+  ctx: AppContext,
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  next: () => Promise<any>
+): Promise<void> {
+  const sig = ctx.headers['signature']
+  const sigInput = ctx.headers['signature-input']
+
+  if (
+    !sig ||
+    !sigInput ||
+    typeof sig !== 'string' ||
+    typeof sigInput !== 'string'
+  ) {
+    ctx.status = 400
+    ctx.body = {
+      error: 'invalid_request',
+      message: 'invalid signature headers'
+    }
+    next()
+    return
+  }
+
+  const { body } = ctx.request
+  const { path, method } = ctx
+  let verified: VerifySigResult
+  if (
+    path.includes('/introspect') &&
+    method === HttpMethod.POST.toUpperCase()
+  ) {
+    const accessToken = await AccessToken.query().findOne(
+      'value',
+      body['access_token']
+    )
+    if (!accessToken) {
+      ctx.status = 401
+      ctx.body = {
+        error: 'invalid_client',
+        message: 'invalid access token'
+      }
+
+      return
+    }
+
+    const grant = await Grant.query().findById(accessToken.grantId)
+    verified = await verifySigFromBoundKey(deps, sig, sigInput, grant, ctx)
+  } else if (
+    path.includes('/token') &&
+    method === HttpMethod.DELETE.toUpperCase()
+  ) {
+    const accessToken = await AccessToken.query().findOne(
+      'managementId',
+      ctx.params['managementId']
+    )
+    if (!accessToken) {
+      ctx.status = 401
+      ctx.body = {
+        error: 'invalid_client',
+        message: 'invalid access token'
+      }
+      return
+    }
+
+    const grant = await Grant.query().findById(accessToken.grantId)
+    verified = await verifySigFromBoundKey(deps, sig, sigInput, grant, ctx)
+  } else if (path.includes('/continue')) {
+    const grant = await Grant.query().findOne(
+      'interactId',
+      ctx.params['interactId']
+    )
+    if (!grant) {
+      ctx.status = 401
+      ctx.body = {
+        error: 'invalid_interaction',
+        message: 'invalid grant'
+      }
```

Comment:
@njlie we should not be doing any direct model querying outside of services. These should be access methods on the service and then be using the service method for them. It makes maintaining it far easier.

Reply:
Is this method disqualified from direct model querying because it's designated as a middleware, or because it crosses the boundary between models and their corresponding services (i.e. it's a `client` service calling into a `Grant` model)?

Reply:
The later, the models should really be reserved for usage within the service or tests when really required.